### ISSUE TITLE
[no-null-keyword] add no-null-keyword rule

### ISSUE THREAD

**Issue by a-tarasyuk** [typescript-eslint/typescript-eslint#676]
In the [ROADMAP](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/ROADMAP.md) there is mention to use [eslint-plugin-no-null](https://github.com/nene/eslint-plugin-no-null) instead of `no-null-keyword`, however, `no-null/no-null` doesn't handle `null` type. 

Is there any plan to implement this `no-null-keyword` as part of `@typescript-eslint/eslint-plugin`?

**Comment by bradzacher** [comment]
Use the `ban-types` rule.

**Comment by a-tarasyuk** [comment]
@bradzacher

```JSON
"@typescript-eslint/ban-types": ["error", {
  "types": {
    "null": "Use 'undefined' instead of 'null'"
  }
}],
```
Is it right configuration?
Is it possible to use `ban-types` only for `null` types?

**Comment by bradzacher** [comment]
actually I'm wrong. I forgot that `null` is treated as a keyword (parses to `TSNullKeyword`, not TSTypeReference`) so `ban-types` won't work out of the box.

It should probably be added to rule.
I don't like the idea of adding new rules that do the same things.

There is `no-explicit-any`, but that rule exists for a common case of making code stricter, where as a `no-null-keyword` rule is a pure style thing which could be covered by `ban-types`.

TBH I don't like the premise of the `no-null-keyword` rule either.
Instead of banning the `null`, you should just adjust your equality checks (use `== null` instead of `=== null/undefined` and you'll handle both cases).

**Comment by a-tarasyuk** [comment]
@bradzacher 
> Is it possible to use ban-types only for null types?

What about this case?
 
> I don't like the idea of adding new rules that do the same things.

Ok, maybe make sense to create an external plugin for this rule?

**Comment by bradzacher** [comment]
No, it can be added to this rule.
I don't ever block rules / options based on my stylistic preferences unless I think they are completely backwards.

I'm happy for there to be an option added to `ban-types` for null/undefined types.

---

> Is it possible to use ban-types only for null types?

See:

> `null` is treated as a keyword (parses to `TSNullKeyword`, not `TSTypeReference`)

**Comment by a-tarasyuk** [comment]
> See:

@bradzacher Sorry, I mean the following case.

When a user enables @typescript-eslint/ban-types, then the following types will be handled

```TS
defaultOptions: [{
  types: {
    String: {
      message: 'Use string instead',
      fixWith: 'string',
    },
    Boolean: {
      message: 'Use boolean instead',
      fixWith: 'boolean',
    },
    Number: {
      message: 'Use number instead',
      fixWith: 'number',
    },
    Object: {
      message: 'Use Record<string, any> instead',
      fixWith: 'Record<string, any>',
    },
    Symbol: {
      message: 'Use symbol instead',
      fixWith: 'symbol',
    },
  },
}]

When a user adds custom type 
```
"@typescript-eslint/ban-types": ["error", {
  "types": {
    "null": "Use 'undefined' instead of 'null'"
  }
}]

I think options will be merged 

```TS
types: {
    String: {
      message: 'Use string instead',
      fixWith: 'string',
    },
    Boolean: {
      message: 'Use boolean instead',
      fixWith: 'boolean',
    },
    Number: {
      message: 'Use number instead',
      fixWith: 'number',
    },
    Object: {
      message: 'Use Record<string, any> instead',
      fixWith: 'Record<string, any>',
    },
    Symbol: {
      message: 'Use symbol instead',
      fixWith: 'symbol',
    },
   "null": "Use 'undefined' instead of 'null'"
}
```

Or I'm wrong? Is is poosible to enable only "null"? Or "null" and "String"?

**Comment by bradzacher** [comment]
The default options are overridden by your settings, they do not get merged.

**Comment by a-tarasyuk** [comment]
@bradzacher 

I see that for this configuration

```json
"@typescript-eslint/ban-types": ["error", {
  "types": {
    "null": "Use 'undefined' instead of 'null'"
  }
}]
```

@typescript-eslint/ban-types receives the following options in `bannedTyped` variable

 ```Shell
{ String: { message: 'Use string instead', fixWith: 'string' },
  Boolean: { message: 'Use boolean instead', fixWith: 'boolean' },
  Number: { message: 'Use number instead', fixWith: 'number' },
  Object:
   { message: 'Use Record<string, any> instead',
     fixWith: 'Record<string, any>' },
  Symbol: { message: 'Use symbol instead', fixWith: 'symbol' },
  null: 'Error' }
```

**Comment by bradzacher** [comment]
huh... that's interesting - a weird side-effect of the default options code... Technically a bug.

**Comment by a-tarasyuk** [comment]
👍  I think the problem in [applyDefault](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/experimental-utils/src/eslint-utils/RuleCreator.ts#L46) -> [deepMerge](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/experimental-utils/src/eslint-utils/applyDefault.ts#L26)

**Comment by a-tarasyuk** [comment]
I've checked TSLint `ban-types`, it doesn't handle `NullKeyword`, and I think doesn't make sense to add `TSNullKeyword`, because `no-null-keyword` behaves differently even we add `TSNullKeyword`. For instance, the following code is invalid code for TSLint rule `no-null-keyword`

```TS
const a: null;
const b = null;
interface A {
   a: null
}
function a(b: null) {}
```

In the same time, it allows 

```TS
type X = null | undefined;
```

After `@typescript-eslint/ban-types` will handle `TSNullKeyword` it will throw an error for the for following cases 

```TS
const a: null;
interface A {
   a: null
}
function a(b: null) {}
type X = null | undefined; // <- Error, however, `no-null-keyword` ignores it
```
and [`eslint-plugin-no-null`](https://github.com/nene/eslint-plugin-no-null) should throw an error for `Literal`s

```TS
const b = null;
a(null);
```

**Comment by uniqueiniquity** [comment]
From looking at the tslint source for `no-null-keyword`, it looks like it's actually a bug that it gives errors on `const a: null`. Based on their tests, as well as the way the rule is implemented, I believe the intent is that it shouldn't look at type positions at all. Therefore, it's correct that `eslint-plugin/no-null` implements the _intended_ behavior of tslint's `no-null-keyword`.
